README.md
# hyper*x*

[![Rustdoc](https://docs.rs/hyperx/badge.svg)](https://docs.rs/hyperx)
[![Change Log](https://img.shields.io/crates/v/hyperx.svg?maxAge=3600&label=change%20log&color=9cf)](https://github.com/dekellum/hyperx/blob/main/CHANGELOG.md)
[![crates.io](https://img.shields.io/crates/v/hyperx.svg?maxAge=3600)](https://crates.io/crates/hyperx)
[![CI Status](https://github.com/dekellum/hyperx/workflows/CI/badge.svg?branch=main)](https://github.com/dekellum/hyperx/actions?query=workflow%3ACI)
[![deps status](https://deps.rs/repo/github/dekellum/hyperx/status.svg)](https://deps.rs/repo/github/dekellum/hyperx)

[Hyper] is the low-level HTTP implementation for Rust. Hyper*x* is an
e*x*traction of the hyper 0.11 typed header module, with minimized
dependencies, for continued use with hyper 0.12 or later,
where it was dropped.

[Hyper]: https://github.com/hyperium/hyper

## Minimum supported rust version

MSRV := 1.46.0

The crate will fail fast on any lower rustc (via a build.rs version check) and
is also CI tested on this version. MSRV will only be increased in a new MINOR
(or MAJOR) release of this crate. However, some direct or transitive
dependencies unfortunately have or may increase MSRV in PATCH releases. Users
may need to selectively control updates by preserving/distributing a Cargo.lock
file in order to control MSRV.

## License

The MIT license ([LICENSE](LICENSE) or http://opensource.org/licenses/MIT)


CHANGELOG.md
## 1.4.0 (2021-10-29)

* Minimum supported rust version (MSRV) is now 1.46.0 (_http_ 0.2.5 MSRV is
  also 1.46.0).

* Upgrade to language-tags 0.3.1. (#35)

* Remove _httparse_ dependency and conversion from `httparse::Error` to
  `hyperx::Error`.  In the presumed unlikely event that this conversion is used
  externally, this is a potential breaking change, risked in a minor
  release. (#32)

* Broaden _httpdate_ dependency to include the 1.0 major release.

* Add docs to cookie headers recommending use of the _cookie_ crate. (#30)

## 1.3.1 (2021-10-28)

* Broaden _bytes_ dependency to include 1.1.

* Broaden _httparse_ dependency to include 1.4.

* Clarify MSRV policy in README.md.

## 1.3.0 (2021-1-8)

* Remove _log_ dependency, which was only used sparsely as a poor workaround
  for unspecific `Error::Header` errors. ([#25])

* Upgrade to _bytes_ 1.0.0 (paolobarbolini [#31])

* Lift prior constraint on _http_ dependency (optimistically hoping for no
  further 0.2.z anomalous releases).

[#25]: https://github.com/dekellum/hyperx/pull/25
[#31]: https://github.com/dekellum/hyperx/pull/31

## 1.2.1 (2021-1-7)

* Constrain _http_ dependency to <0.2.3 for remaining 1.2 series due to dubious
  release practices, e.g. forcing duplicates in patch releases. This will be
  lifted in 1.3.0, see this [github comment][461].

[461]: https://github.com/hyperium/http/pull/461#issuecomment-756298944

## 1.2.0 (2020-10-6)

* Replace use of time crate with httpdate crate for date/time typed
  headers (paolobarbolini #24)

* Broaden _base64_ dependency to include 0.13.0 (dvermd #23)

## 1.1.0 (2020-8-29)

* Fix various compile warnings with rustc 1.44 and later.

* Broaden _base64_ dependency to include 0.12.0 (paolobarbolini #20)

## 1.0.0 (2020-1-3)

* The default _compat_ feature is no longer optional, as its unlikely that
  anyone would be using recent versions without this feature. The feature gate
  name is retained for now, but has no effect.

* Place the legacy `Headers` struct under a new non-default _headers_ feature
  gate. Note that use of this type is no longer required nor recommended for
  parsing and serialization of typed headers. See the rewritten typed header
  doc examples.  Consider replacing with the _http_ crate `HeaderMap` and the
  `TypedHeaders` extension trait introduced here in 0.15.0. (#18)

* Upgrade to _http_ 0.2.0 (\w API changes) and _bytes_ 0.5.2 (MSRV 1.39.0)

* Upgrade (unconstrain) _cfg-if_ dependency to 0.1.10 (MSRV 1.31.0)

* Upgrade to _unicase_ 2.6.0

* Upgrade to _percent-encoding_ 2.1.0 (\w API changes, MSRV 1.33.0) (#15)

* Upgrade to _time_ 0.1.39 to avoid minimal version build failure

* Broaden _base64_ dependency to include 0.11.0 (MSRV 1.34.0)

* MSRV is now 1.39.0, based on above upgrades.

## 0.15.2 (2019-10-1)

* Constrain transitive _cfg-if_ dependency to <0.1.10 to preserve MSRV 1.27.2.

* Narrow various other dependencies for future reliability.  We may
  subsequently make PATCH releases which _broaden_ private or public
  dependencies to include new releases found compatible.

## 0.15.1 (2019-6-3)

* Fix build.rs for `rustc --version` not including git metadata (alyssais #14)

## 0.15.0 (2019-5-8)

* Add a `TypedHeaders` extension trait providing more convenient generic
  encode/decode methods to `http::HeaderMap` for _hyperx_ typed headers,
  implemented using a new `StandardHeader` trait and `standard_header!` macro,
  with an associate function for the `HeaderName` constants of the _http_
  crate. (#13)

* Add reference based `impl From<&'a Headers> for http::HeaderMap` for symmetry
  and performance, e.g. avoiding a `clone`. (#13)

* Increase MSRV to 1.27.2, which enables us to revert a CI workaround for the
  fact that base64 0.10.1 was released with this same MSRV. (#10 #12)

* Add a build.rs to check MSRV and fail fast with a clear error when older
  rustc versions are used. (#12)

## 0.14.0 (2019-1-4)

* Update the signature of `Header::parse_header` to be generic over types
  implementing a new `RawLike` trait, which includes the existing local `Raw`
  type as well as _http_ crate types `HeaderValue` and (`HeaderMap::get_all`)
  `GetAll`. This avoids an allocation when directly parsing from these later
  types.

  _Expected Breakage_: Any 3rd-party custom headers directly implementing
  `parse_header` will need to change accordingly on upgrade. Also `Into`
  conversions to `Raw` now frequently need to be type annotated. (#8)

* Improve header module rustdoc, including with parsing usage for the above.

## 0.13.2 (2019-1-2)

* Remove un-exported, and unused as of 0.13.1, `uri` module and related code.

* Broaden base64 dependency to include 0.10.0, passing tests.

* Silence a deprecation warning for `str::trim_right_matches` until the minimum
  rust version is updated to 1.30.0.

## 0.13.1 (2018-6-26)

* Remove `error::UriError` re-export and `error::Canceled` which are unused
  internally and where not exported from this crate. (#5)

## 0.13.0 (2018-6-18)

* Remove variants from `hyperx::Error` which are unused by the header
  module. Exhaustive matching has been discouraged for this enum, but this is
  still a potential breaking change. (dekellum #2)

* Add an alternative, by reference `From<&http::HeaderMap>` for `Headers`.
  (DarrenTsung #3)

## 0.12.0 (2018-6-1)

Forked from hyper 0.11.27, e*x*tracting the typed *header* module
from [hyperium/hyper@76fdbcf2], 0.11.x branch, preserved here as
[76fdbcf2].

## Prior Releases

See [hyper's CHANGELOG] for prior updates pertaining to the headers
sub-module.

[hyper's CHANGELOG]: https://github.com/hyperium/hyper/blob/0.11.x/CHANGELOG.md
[hyperium/hyper@76fdbcf2]: https://github.com/hyperium/hyper/commit/76fdbcf2

src/header/compat.rs
//! Implementation module for various compatibility features with the _http_
//! crate.

use std::fmt::Display;

use http::header::{GetAll, HeaderMap, HeaderValue, ValueIter};

use ::Result;
use super::{Header, RawLike};

#[cfg(feature = "headers")]
use std::convert::From;

#[cfg(feature = "headers")]
use http;

#[cfg(feature = "headers")]
use super::{Raw, Headers};

/// A trait for the "standard" headers that have an associated `HeaderName`
/// constant in the _http_ crate.
pub trait StandardHeader: Header + Sized {
    /// The `HeaderName` from the _http_ crate for this header.
    fn http_header_name() -> ::http::header::HeaderName;
}

/// Extension trait for `decode` (parsing) and `encode` (serialization) of
/// typed headers from/to a collection of headers such as `http::HeaderMap`.
pub trait TypedHeaders {
    /// Decode and return `Header` type H or `Error::Header`.
    ///
    /// `Error::Header` is returned on failed parse, or for a single-valued
    /// Header type, if no values or multiple values are found in the
    /// collection.  Multi-valued header types such as `ContentEncoding` will
    /// instead return an empty list value if no values are found.  To
    /// distinguish the not found case, use `try_decode` instead.
    fn decode<H>(&self) -> Result<H>
        where H: StandardHeader;

    /// Decode and return `Header` type H or `Error::Header` if found, or
    /// return `None` if not found.
    ///
    /// This variant will return `Option::None` if no header with the
    /// associated key (`HeaderName`) is found in the collection. If the
    /// collection does contain such a key, it will return the header type H or
    /// `Error::Header`.
    fn try_decode<H>(&self) -> Option<Result<H>>
        where H: StandardHeader;

    /// Encode and write the specified typed header value in the collection.
    ///
    /// Uses the `Display` format of the provided header value to write a single
    /// header. This will overwrite any preexisting values with the same
    /// key (`HeaderName`). Use `encode_append` instead to avoid this.
    fn encode<H>(&mut self, value: &H)
        where H: StandardHeader + Display;

    /// Encode and append the specified typed header value into the collection.
    ///
    /// Uses the `Display` format of the provided header value to append a
    /// single header. If the collection previously had a value for the same
    /// key, the additional value is appended to the end.
    fn encode_append<H>(&mut self, value: &H)
        where H: StandardHeader + Display;
}

/// Iterator adaptor for HeaderValue
#[derive(Debug)]
pub struct ValueMapIter<'a>(ValueIter<'a, HeaderValue>);

impl TypedHeaders for HeaderMap {
    fn decode<H>(&self) -> Result<H>
        where H: StandardHeader
    {
        let vals = self.get_all(H::http_header_name());
        H::parse_header(&vals)
    }

    fn try_decode<H>(&self) -> Option<Result<H>>
        where H: StandardHeader
    {
        let hname = H::http_header_name();
        if self.contains_key(&hname) {
            let vals = self.get_all(&hname);
            Some(H::parse_header(&vals))
        } else {
            None
        }
    }

    fn encode<H>(&mut self, val: &H)
        where H: StandardHeader + Display
    {
        self.insert(
            H::http_header_name(),
            val.to_string().parse().expect("header value"));
    }

    fn encode_append<H>(&mut self, val: &H)
        where H: StandardHeader + Display
    {
        self.append(
            H::http_header_name(),
            val.to_string().parse().expect("header value"));
    }
}

#[cfg(feature = "headers")]
impl From<http::HeaderMap> for Headers {
    fn from(mut header_map: http::HeaderMap) -> Headers {
        let mut headers = Headers::new();
        for (name, mut value_drain) in header_map.drain() {
            if let Some(first_value) = value_drain.next() {
                let mut raw: Raw = first_value.as_bytes().into();
                for value in value_drain {
                    raw.push(value.as_bytes());
                }
                headers.append_raw(name.as_str().to_string(), raw);
            }
        }
        headers
    }
}

#[cfg(feature = "headers")]
impl<'a> From<&'a http::HeaderMap> for Headers {
    fn from(header_map: &'a http::HeaderMap) -> Headers {
        let mut headers = Headers::new();
        for (name, value) in header_map.iter() {
            headers.append_raw_str(name.as_str(), value.as_bytes());
        }
        headers
    }
}

#[cfg(feature = "headers")]
impl From<Headers> for http::HeaderMap {
    #[inline]
    fn from(headers: Headers) -> http::HeaderMap {
        http::HeaderMap::from(&headers)
    }
}

#[cfg(feature = "headers")]
impl<'a> From<&'a Headers> for http::HeaderMap {
    fn from(headers: &'a Headers) -> http::HeaderMap {
        let mut hmap = http::HeaderMap::new();
        for header in headers.iter() {
            let entry = hmap.entry(header.name())
                .expect("convert invalid header name");
            let mut value_iter = header.raw().iter().map(|line| {
                http::header::HeaderValue::from_bytes(line)
                    .expect("convert invalid header value")
            });
            match entry {
                http::header::Entry::Occupied(mut  occupied) => {
                    for value in value_iter {
                        occupied.append(value);
                    }
                },
                http::header::Entry::Vacant(vacant) => {
                    if let Some(first_value) = value_iter.next() {
                        let mut occupied = vacant.insert_entry(first_value);
                        for value in value_iter {
                            occupied.append(value);
                        }
                    }
                }
            }
        }
        hmap
    }
}

impl<'a> Iterator for ValueMapIter<'a> {
    type Item = &'a [u8];

    fn next(&mut self) -> Option<Self::Item> {
        self.0.next().map(HeaderValue::as_bytes)
    }
}

impl<'a> RawLike<'a> for GetAll<'a, HeaderValue> {
    type IntoIter = ValueMapIter<'a>;

    fn len(&'a self) -> usize {
        self.iter().count()
    }

    fn one(&'a self) -> Option<&'a [u8]> {
        let mut iter = self.iter();
        if let Some(v) = iter.next() {
            if iter.next().is_none() {
                return Some(v.as_bytes());
            }
        }
        None
    }

    fn iter(&'a self) -> ValueMapIter<'a> {
        ValueMapIter(self.iter())
    }
}

impl<'a> RawLike<'a> for &'a HeaderValue {
    type IntoIter = ::std::iter::Once<&'a [u8]>;

    fn len(&'a self) -> usize {
        1
    }

    fn one(&'a self) -> Option<&'a [u8]> {
        Some(self.as_bytes())
    }

    fn iter(&'a self) -> Self::IntoIter {
        ::std::iter::once(self.as_bytes())
    }
}

#[cfg(test)]
mod tests {
    use http;
    use ::header::{
        ContentEncoding, ContentLength, Encoding, ETag,
        Header, Te, TypedHeaders};

    #[cfg(feature = "headers")]
    use ::header::{Headers, Host};

    #[cfg(feature = "nightly")]
    use test::Bencher;

    #[cfg(feature = "nightly")]
    use ::header::EntityTag;

    #[test]
    fn test_empty_decode() {
        let hmap = http::HeaderMap::new();
        let len = hmap.decode::<ContentLength>();
        assert!(len.is_err());
    }

    #[test]
    fn test_empty_decode_etag() {
        let hmap = http::HeaderMap::new();
        let len = hmap.decode::<ETag>();
        assert!(len.is_err());
    }

    #[test]
    fn test_empty_decode_te() {
        let hmap = http::HeaderMap::new();
        let te = hmap.decode::<Te>().unwrap();
        assert_eq!(te, Te(vec![]));
    }

    #[test]
    fn test_empty_decode_content_encoding() {
        let hmap = http::HeaderMap::new();
        let ce = hmap.decode::<ContentEncoding>().unwrap();
        assert_eq!(ce, ContentEncoding(vec![]));
    }

    #[test]
    fn test_empty_try_decode() {
        let hmap = http::HeaderMap::new();
        let len = hmap.try_decode::<ContentLength>();
        assert!(len.is_none());
    }

    #[test]
    fn test_empty_try_decode_te() {
        let hmap = http::HeaderMap::new();
        let te = hmap.try_decode::<Te>();
        assert!(te.is_none());
    }

    #[test]
    fn test_decode() {
        let mut hmap = http::HeaderMap::new();
        hmap.insert(http::header::CONTENT_LENGTH, "11".parse().unwrap());
        let len: ContentLength = hmap.decode().unwrap();
        assert_eq!(*len, 11);
    }

    #[test]
    fn test_encode_decode() {
        let mut hmap = http::HeaderMap::new();
        hmap.encode(&ContentLength(11));
        let len: ContentLength = hmap.decode().unwrap();
        assert_eq!(*len, 11);
    }

    #[test]
    fn test_empty_encode() {
        let mut hmap = http::HeaderMap::new();
        hmap.encode(&ContentEncoding(vec![]));
        assert_eq!(hmap.len(), 1);
        let ce: ContentEncoding = hmap.decode().unwrap();
        assert_eq!(*ce, vec![]);
    }

    #[test]
    fn test_empty_encode_2() {
        let mut hmap = http::HeaderMap::new();
        hmap.encode(&ContentEncoding(vec![]));
        hmap.encode_append(&ContentEncoding(vec![]));
        assert_eq!(hmap.len(), 2);
        let ce: ContentEncoding = hmap.decode().unwrap();
        assert_eq!(*ce, vec![]);
    }

    #[test]
    fn test_encode_append() {
        let mut hmap = http::HeaderMap::new();
        hmap.encode_append(
            &ContentEncoding(vec![Encoding::Identity]));
        hmap.encode_append(
            &ContentEncoding(vec![Encoding::Gzip, Encoding::Chunked]));
        let ce: ContentEncoding = hmap.decode().unwrap();
        assert_eq!(
            *ce,
            vec![Encoding::Identity, Encoding::Gzip, Encoding::Chunked]);
    }

    #[cfg(feature = "headers")]
    fn raw_headers_sample() -> Headers {
        let mut heads = Headers::new();

        heads.set_raw("date", b"Thu, 02 May 2019 21:10:03 GMT".as_ref());
        heads.set_raw("connection", b"Keep-Alive".as_ref());
        heads.set_raw("accept-ranges", b"bytes".as_ref());
        heads.set_raw("etag", b"\"1544639720\"".as_ref());
        heads.set_raw("transfer-encoding", b"gzip".as_ref());
        heads.append_raw("transfer-encoding", b"chunked".as_ref());
        heads.set_raw("content-length", b"7050".as_ref());
        heads.set_raw("content-type", b"text/css; charset=utf-8".as_ref());
        heads.set_raw("last-modified", b"Wed, 12 Dec 2018 18:35:20 GMT".as_ref());
        heads.set_raw("x-hello-human",
                      b"Say hello back! @getBootstrapCDN on Twitter".as_ref());
        heads.set_raw("access-control-allow-origin", b"*".as_ref());
        heads.set_raw("vary", b"Accept-Encoding".as_ref());
        heads.set_raw("x-cache", b"HIT".as_ref());
        heads.set_raw("timing-allow-origin", b"*".as_ref());
        heads.set_raw("cache-control", b"public, max-age=31536000".as_ref());

        heads
    }

    #[cfg(feature = "headers")]
    #[test]
    fn test_convert_mixed() {
        let mut headers = Headers::new();
        headers.set(ContentLength(11));
        headers.set(Host::new("foo.bar", None));
        headers.append_raw("x-foo", b"bar".to_vec());
        headers.append_raw("x-foo", b"quux".to_vec());

        let mut hmap = http::HeaderMap::new();
        hmap.insert(http::header::CONTENT_LENGTH, "11".parse().unwrap());
        hmap.insert(http::header::HOST, "foo.bar".parse().unwrap());
        hmap.append("x-foo", "bar".parse().unwrap());
        hmap.append("x-foo", "quux".parse().unwrap());

        let headers2: Headers = hmap.clone().into();
        let hmap2: http::HeaderMap = headers.clone().into();
        assert_eq!(headers, headers2);
        assert_eq!(headers2.len(), 3);
        assert_eq!(hmap, hmap2);
        assert_eq!(hmap2.len(), 4);
    }

    #[test]
    #[cfg(feature = "headers")]
    fn test_convert_sample() {
        let headers = raw_headers_sample();
        let hmap = http::HeaderMap::from(headers.clone());
        let headers2 = Headers::from(hmap.clone());
        let hmap2 = http::HeaderMap::from(headers2.clone());
        assert_eq!(headers, headers2);
        assert_eq!(headers2.len(), 14);
        assert_eq!(hmap2, hmap);
        assert_eq!(hmap2.len(), 15);
    }

    #[test]
    #[cfg(feature = "headers")]
    fn test_convert_by_ref() {
        let headers = raw_headers_sample();
        let hmap = http::HeaderMap::from(&headers);
        let headers2 = Headers::from(&hmap);
        let hmap2 = http::HeaderMap::from(&headers2);
        assert_eq!(headers, headers2);
        assert_eq!(headers2.len(), 14);
        assert_eq!(hmap2, hmap);
        assert_eq!(hmap2.len(), 15);
    }

    #[test]
    fn test_value_parse() {
        let mut hmap = http::HeaderMap::new();
        hmap.insert(http::header::CONTENT_ENCODING,
                    "chunked, gzip".parse().unwrap());
        let val = hmap.get(http::header::CONTENT_ENCODING).unwrap();
        let ce = ContentEncoding::parse_header(&val).unwrap();
        assert_eq!(ce, ContentEncoding(vec![Encoding::Chunked, Encoding::Gzip]))
    }

    #[test]
    fn test_multi_value_parse() {
        let mut hmap = http::HeaderMap::new();
        hmap.insert(http::header::CONTENT_ENCODING,
                    "chunked, gzip".parse().unwrap());
        hmap.append(http::header::CONTENT_ENCODING,
                    "br".parse().unwrap());

        let vals = hmap.get_all(http::header::CONTENT_ENCODING);
        let ce = ContentEncoding::parse_header(&vals).unwrap();
        assert_eq!(
            ce,
            ContentEncoding(vec![
                Encoding::Chunked, Encoding::Gzip, Encoding::Brotli
            ])
        )
    }

    #[cfg(feature = "nightly")]
    #[bench]
    fn bench_0_value_parse(b: &mut Bencher) {
        let mut hmap = http::HeaderMap::new();
        hmap.insert(http::header::CONTENT_ENCODING,
                    "chunked, gzip".parse().unwrap());
        b.iter(|| {
            let val = hmap.get(http::header::CONTENT_ENCODING).unwrap();
            ContentEncoding::parse_header(&val).unwrap();
        })
    }

    #[cfg(feature = "nightly")]
    #[bench]
    fn bench_0_value_parse_extra_str(b: &mut Bencher) {
        use header::Raw;
        let mut hmap = http::HeaderMap::new();
        hmap.insert(http::header::CONTENT_ENCODING,
                    "chunked, gzip".parse().unwrap());
        b.iter(|| {
            let val = hmap.get(http::header::CONTENT_ENCODING).unwrap();
            let r: Raw = val.to_str().unwrap().into();
            ContentEncoding::parse_header(&r).unwrap();
        })
    }

    #[cfg(feature = "nightly")]
    #[bench]
    fn bench_0_value_parse_int(b: &mut Bencher) {
        let mut hmap = http::HeaderMap::new();
        hmap.insert(http::header::CONTENT_LENGTH, "1024".parse().unwrap());
        b.iter(|| {
            let val = hmap.get(http::header::CONTENT_LENGTH).unwrap();
            ContentLength::parse_header(&val).unwrap();
        })
    }

    #[cfg(feature = "nightly")]
    #[bench]
    fn bench_1_get_parse_int(b: &mut Bencher) {
        let mut hmap = http::HeaderMap::new();
        hmap.insert(http::header::CONTENT_LENGTH, "11".parse().unwrap());
        b.iter(|| {
            let vals = hmap.get_all(http::header::CONTENT_LENGTH);
            let len = ContentLength::parse_header(&vals).unwrap();
            assert_eq!(*len, 11);
        })
    }

    #[cfg(feature = "nightly")]
    #[bench]
    fn bench_1_get_parse_int_one(b: &mut Bencher) {
        let mut hmap = http::HeaderMap::new();
        hmap.insert(http::header::CONTENT_LENGTH, "11".parse().unwrap());
        b.iter(|| {
            let val = hmap.get(http::header::CONTENT_LENGTH).unwrap();
            let len = ContentLength::parse_header(&val).unwrap();
            assert_eq!(*len, 11);
        })
    }

    #[cfg(feature = "nightly")]
    #[bench]
    fn bench_2_decode_int(b: &mut Bencher) {
        let mut hmap = http::HeaderMap::new();
        hmap.insert(http::header::CONTENT_LENGTH, "11".parse().unwrap());
        b.iter(|| {
            let len: ContentLength = hmap.decode().unwrap();
            assert_eq!(*len, 11);
        })
    }

    #[cfg(feature = "nightly")]
    #[bench]
    fn bench_2_try_decode_int(b: &mut Bencher) {
        let mut hmap = http::HeaderMap::new();
        hmap.insert(http::header::CONTENT_LENGTH, "11".parse().unwrap());
        b.iter(|| {
            let len: ContentLength = hmap.try_decode().unwrap().unwrap();
            assert_eq!(*len, 11);
        })
    }

    #[cfg(all(feature = "nightly", feature = "headers"))]
    #[bench]
    fn bench_3_get_orig_int(b: &mut Bencher) {
        let mut hdrs = ::header::Headers::new();
        hdrs.set_raw("content-length", "11");
        b.iter(|| {
            let len: &ContentLength = hdrs.get().unwrap();
            assert_eq!(**len, 11);
        })
    }

    #[cfg(feature = "nightly")]
    #[bench]
    fn bench_4_encode_int(b: &mut Bencher) {
        b.iter(|| {
            let mut hmap = http::HeaderMap::new();
            hmap.encode(&ContentLength(11));
            assert_eq!(hmap.len(), 1);
        })
    }

    #[cfg(feature = "nightly")]
    #[bench]
    fn bench_4_encode_multi(b: &mut Bencher) {
        b.iter(|| {
            let mut hmap = http::HeaderMap::new();
            hmap.encode(
                &ContentEncoding(vec![Encoding::Identity]));
            hmap.encode_append(
                &ContentEncoding(vec![Encoding::Gzip, Encoding::Chunked]));
            hmap.encode(&ContentLength(11));
            hmap.encode(
                &ETag(EntityTag::strong("pMMV3zmCrXr-n4ZZLR9".to_owned())));
            assert_eq!(hmap.len(), 4);
        })
    }

    #[cfg(all(feature = "nightly", feature = "headers"))]
    #[bench]
    fn bench_5_map_from_headers(b: &mut Bencher) {
        let heads = raw_headers_sample();
        b.iter(|| {
            let hmap = http::HeaderMap::from(&heads);
            assert_eq!(hmap.len(), 15);
        })
    }

    #[cfg(all(feature = "nightly", feature = "headers"))]
    #[bench]
    fn bench_5_headers_from_map(b: &mut Bencher) {
        let heads = raw_headers_sample();
        let hmap: http::HeaderMap = heads.into();
        b.iter(|| {
            let heads = Headers::from(&hmap);
            assert_eq!(heads.len(), 14);
        })
    }

    #[cfg(all(feature = "nightly", feature = "headers"))]
    #[bench]
    fn bench_5_headers_from_map_by_value(b: &mut Bencher) {
        let heads = raw_headers_sample();
        let hmap: http::HeaderMap = heads.into();
        b.iter(|| {
            let heads = Headers::from(hmap.clone());
            assert_eq!(heads.len(), 14);
        })
    }
}
